src/breakpoints.jl

```diff
@@ -1,26 +1,104 @@
-    if idx === nothing
-        push!(_breakpoints, bp)
+function framecode_matches_breakpoint(framecode::FrameCode, bp::BreakpointSignature)
+    function extract_function_from_method(m::Method)
+        sig = Base.unwrap_unionall(m.sig)
+        ft0 = sig.parameters[1]
+        ft = Base.unwrap_unionall(ft0)
+        if ft <: Function && isa(ft, DataType) && isdefined(ft, :instance)
+            return ft.instance
+        elseif isa(ft, DataType) && ft.name === Type.body.name
+            f = ft.parameters[1]
+        else
+            return ft
+        end
     end
+
+    framecode.scope isa Method || return false
+    meth = framecode.scope
+    bp.f isa Method && return meth === bp.f
+    bp.f === extract_function_from_method(meth) || return false
+    bp.sig === nothing && return true
+    return bp.sig <: meth.sig
+end
+
+"""
+    breakpoint(f, [sig], [line], [condition])
+
+Add a breakpoint to `f` with the specified argument types `sig`.¨
+If `sig` is not given, the breakpoint will apply to all methods of `f`.
+If `f` is a method, the breakpoint will only apply to that method.
+Optionally specify an absolute line number `line` in the source file; the default
+is to break upon entry at the first line of the body.
+Without `condition`, the breakpoint will be triggered every time it is encountered;
+the second only if `condition` evaluates to `true`.
+`condition` should be written in terms of the arguments and local variables of `f`.
+
+# Example
+```julia
+function radius2(x, y)
+    return x^2 + y^2
+end
+
+breakpoint(radius2, Tuple{Int,Int}, :(y > x))
+```
+"""
+function breakpoint(f, sig=nothing, line::Integer=0, condition::Condition=nothing)
+    sig !== nothing && (sig = Base.to_tuple_type(sig))
+    bp = BreakpointSignature(f, sig, line, condition, Ref(true), BreakpointRef[])
+    add_to_existing_framecodes(bp)
+    idx = findfirst(bp2 -> same_location(bp, bp2), _breakpoints)
+    idx === nothing ? push!(_breakpoints, bp) : (_breakpoints[idx] = bp)
+    return bp
+end
+breakpoint(f, sig, condition::Condition) = breakpoint(f, sig, 0, condition)
+breakpoint(f, line::Integer, condition::Condition=nothing) = breakpoint(f, nothing, line, condition)
+breakpoint(f, condition::Condition) = breakpoint(f, nothing, 0, condition)
+
+
+"""
+    breakpoint(file, line, [condition])
+
+Set a breakpoint in `file` at `line`. The argument `file` can be a filename, a partial path or absolute path.
+For example, `file = foo.jl` will match against all files with the name `foo.jl`,
+`file = src/foo.jl` will match against all paths containing `src/foo.jl`, e.g. both `Foo/src/foo.jl` and `Bar/src/foo.jl`.
+Absolute paths only matches against the file with that exact absolute path.
+"""
+function breakpoint(file::String, line::Integer, condition::Condition=nothing)
+    file = normpath(file)
+    bp = BreakpointFileLocation(file, CodeTracking.maybe_fix_path(abspath(file)), line, condition, Ref(true), BreakpointRef[])
```

Comment:
Perhaps it's deliberate, but this won't work for non-absolute Base paths, e.g., `"numbers.jl"`.

Reply:
What won't work? If we give it as `numbers.jl` it will be stored as `"numbers.jl"` in the `path` field and we will match against all paths that ends with `number.jl`.

The only reason for the `maybe_fix_path` is if someone gives the "buildbot path" to a method by e.g. using `(@which uuid1()).file` (which the tests does).

Reply:
Good point. Perhaps I don't understand the motivation for both `path` and `abspath` in that struct.

Reply:
Yeah, I can understand that is confusing. The reason is if someone adds a path like `"../foo.jl"`.  If we only store `"../foo.jl"` and the user moves directory we will start referring to another file which is weird. However, we can't store the absolute path only because then we cant do the "trailing path matching".

Reply:
I should add a test for this though.

Reply:
And a code comment.